paima-demo/godot-cip-30-frontend/cip_30_callbacks.gd

```diff
@@ -0,0 +1,68 @@
+extends Node
+# If `RefCounted` is used, then GDScript callbacks stop working - they are not called at all.
+# Probably coz references to callbacks are lost (see below).
+# (?) Alternative is to keep reference to `RefCounted` on root node (main.gd)
+
+## Provides CIP-30 compliant functions for Godot wallet.
+## Used to add GDScript callbacks wrapped with JavaScriptBridge into `window.cardano.godot` Object.
+## res://extra-resources/cip-30-paima-shell.html contains JS script that creates `window.cardano.godot` Object
+## and also wraps this callbacks one more time to return Promise result on the JS side
+## as it is not clear how to get `return` value back from JavaScriptBridge callbacks
+## (see https://forum.godotengine.org/t/getting-return-value-from-js-callback/54190/3)
+
+class_name Cip30Callbacks
+var _godot_wallet
+
+# This references must be kept
+# See expample: https://docs.godotengine.org/en/stable/classes/class_javascriptobject.html#javascriptobject
+var _js_cb_get_unused_addresses = JavaScriptBridge.create_callback(_cb_get_unused_addresses)
+var _js_cb_get_used_addresses = JavaScriptBridge.create_callback(_cb_get_used_addresses)
+var _js_cb_sign_data = JavaScriptBridge.create_callback(_cb_sign_data)
+
+func _init(godot_wallet):
+	_godot_wallet = godot_wallet
+
+# TODO: CIP-30 compliant errors
+# CIP-30 callbacks
+## Adding to `window`
+func add_to(window):
+	if !window:
+		print("GD: Browser 'window' not found - skip adding CIP-30 callbacks")
+		return
+	# `window.cardano.godot` Object is created via custom HTML shell and expeted to be not null
+	# see res://extra-resources/cip-30-paima-shell.html
+	window.cardano.godot.callbacks.get_used_addresses = _js_cb_get_used_addresses
+	window.cardano.godot.callbacks.get_unused_addresses = _js_cb_get_unused_addresses
+	window.cardano.godot.callbacks.sign_data = _js_cb_sign_data
+	print("GD: adding CIP-30 callbacks to `window.cardano.godot.callbacks` is done")
+
+func _cb_get_used_addresses(args):
+	prints("GD: _cb_get_used_addresses")
```

Comment:
Logs like here are used to make sure that GDScrpt callbacks are called at all. As some subtle changes that do not follow `JavaScriptBridge` examples exactly can break things and callbacks will stop working silently.